Initial OS/2 build fixes

ydario opened the pull request:
Here is the first set of commits to make QTCreator 4.11 to build and run. While most feature are working, there are still some issues with process spawn (mixed of native and unix code).

ydario committed:
Build system patches for OS/2 platform: use short names and change
paths. Use system sqlite library. Install plugins into bin/plugins dir.

ydario committed:
plugin projectexplorer: disable auto detection of non-OS/2 tool chains. Recognize OS/2 LX binary format and platform triplet.
plugin git: add proper file extension to git executable.
plugin core: file handling.
plugin extensionsystem: recognize OS/2, handle plugin short name, extend BEGINLIBPATH to load plugins.

ydario committed:
Use win process handling model.
Fix environment separator and name handling.
Import files from 2.2 port.
Use windows file handling code for os2 too, use os2 native process
handling.

ydario committed:
Native process handling.

dmik commented on qtcreatordata.pri:
```diff
@@ -29,6 +29,7 @@ defineReplace(stripStaticBase) {
         isEmpty(vcproj):copy2build.variable_out = PRE_TARGETDEPS
         win32:copy2build.commands = $$QMAKE_COPY \"${QMAKE_FILE_IN}\" \"${QMAKE_FILE_OUT}\"
         unix:copy2build.commands = $$QMAKE_COPY ${QMAKE_FILE_IN} ${QMAKE_FILE_OUT}
+        os2:copy2build.commands = $$QMAKE_COPY ${QMAKE_FILE_IN} ${QMAKE_FILE_OUT}
```
I think it makes sense to join this with unix:
```suggestion
        unix|os2:copy2build.commands = $$QMAKE_COPY ${QMAKE_FILE_IN} ${QMAKE_FILE_OUT}
```

dmik commented on src/libs/libs.pro:
```diff
@@ -18,7 +18,7 @@ SUBDIRS   += \
     sqlite
 
-qtHaveModule(quick) {
+!os2:qtHaveModule(quick) {
```
Why is tracing disabled? Is something missing on OS/2?

ydario replied:
tracing code requires OpenGL support, so I disabled it for now.

dmik commented on src/plugins/help/qlitehtml/qlitehtml.pri:
```diff
@@ -142,6 +142,7 @@ exists($$PWD/litehtml/CMakeLists.txt) {
     LIBS += -L$$LITEHTML_LIB_DIR -llitehtml -lgumbo
 
     win32: PRE_TARGETDEPS += $$LITEHTML_LIB_DIR/litehtml.lib $$LITEHTML_LIB_DIR/gumbo.lib
+    else:os2: PRE_TARGETDEPS += $$LITEHTML_LIB_DIR/litehtml.lib $$LITEHTML_LIB_DIR/gumbo.lib
```
```suggestion
    win32|os2: PRE_TARGETDEPS += $$LITEHTML_LIB_DIR/litehtml.lib $$LITEHTML_LIB_DIR/gumbo.lib
```

dmik commented on src/plugins/plugins.pro:
```diff
@@ -71,7 +71,7 @@ qtHaveModule(serialport) {
 }
 
-qtHaveModule(quick) {
+!os2:qtHaveModule(quick) {
```
Why disabled, what's missing? QML for OS/2 is there.

ydario replied:
yes but these modules require tracing code.

dmik commented on src/qtcreatorplugin.pri:
```diff
@@ -95,7 +96,8 @@ include(rpath.pri)
 
 TEMPLATE = lib
-CONFIG += plugin plugin_with_soname
+!os2:CONFIG += plugin plugin_with_soname
+os2:CONFIG += shared dll
```
Thi one shouldn't be necessary. We should check how Windows handles it and do the same.

ydario replied:
windows code uses plugin. The issue could be related to gcc: I'm sure gcc was able to import symbols directly from a .dll while linking, maybe this does no longer work. plugin config does not generate an import lib.

dmik commented on qtcreator.pro:
```diff
@@ -13,6 +13,7 @@ CONFIG   += ordered
 
 SUBDIRS = src share
 unix:!macx:!isEmpty(copydata):SUBDIRS += bin
+os2:!isEmpty(copydata):SUBDIRS += bin
```
```suggestion
unix|os2:!macx:!isEmpty(copydata):SUBDIRS += bin
```

dmik commented on src/shared/qtlockedfile/qtlockedfile.pri:
```diff
@@ -4,6 +4,7 @@ HEADERS += $$PWD/qtlockedfile.h
 SOURCES += $$PWD/qtlockedfile.cpp
 
 unix:SOURCES += $$PWD/qtlockedfile_unix.cpp
+os2:SOURCES += $$PWD/qtlockedfile_unix.cpp
```
```suggestion
unix|os2:SOURCES += $$PWD/qtlockedfile_unix.cpp
```

dmik commented on src/libs/extensionsystem/pluginspec.cpp:
```diff
@@ -209,6 +215,19 @@ QString PluginSpec::name() const
     return d->name;
 }
 
+/*!
+    \fn QString PluginSpec::shortName() const
+    The short plugin name. This is valid after the PluginSpec::Read state is reached.
+
+    \note This property is only available on OS/2.
+*/
+#ifdef  Q_OS_OS2
+QString PluginSpec::shortName() const
+{
+    return d->shortName;
+}
+#endif
+
```
As discussed on Slack, is it really necessary to introduce all this short name burden on the Qt Creator source level? Plugins are dynamically loaded, so we have full control over name resolution. In Qt itself (in QML) I do it like that:
1. Build a plugin DLL with a short name as required by the OS/2 kernel.
2. Create a plugin symlink with a normal (long) name that points to it.
3. `QLibrary` on OS/2 uses `dlopen` instead of `DosLoadModule` which, besides adding fork support, also handles symlink resolution.

I did that this way for QML plugins because they are referenced by full name from `qmldir` files and fixing that would be a major PITA. Looks like Qt Creator is the same case.

Please see https://github.com/bitwiseworks/qtdeclarative-os2/blob/master/src/imports/layouts/layouts.pro as an example. It gives this result in the build tree:
```
drwxr-xr-x 1 root root       0 Mar 17 00:22 .
drwxr-xr-x 1 root root       0 Jan 16 19:10 ..
-rw-r--r-- 1 root root    3689 Sep  2  2019 plugins.qmltypes
-rw-r--r-- 1 root root     125 Sep  2  2019 qmldir
-rw-rw-rw- 1 root root 1518510 Jan 16 19:12 qqlp.dll
-rw-rw-rw- 1 root root 1883200 Mar 17 00:22 qqlpd.dll
lrwxrwxrwx 1 root root       8 Jan 16 19:12 qquicklayoutsplugin.dll -> qqlp.dll
lrwxrwxrwx 1 root root       9 Mar 17 00:22 qquicklayoutsplugind.dll -> qqlpd.dll
```
The "magic" that does symlink creation is dragged in by `include(qml_plugin)` in the .pro file. Look at https://github.com/bitwiseworks/qtbase-os2/blob/master/mkspecs/features/qml_plugin.prf for details (search for "short" and "os2") there.

dmik commented on src/libs/utils/hostosinfo.h:
```diff
@@ -31,7 +31,7 @@
 #include <QString>
 
-#ifdef Q_OS_WIN
+#if defined(Q_OS_WIN) || defined(Q_OS_OS2)
```
Qt for OS/2 has a special define for that — Q_OS_DOSLIKE. It's defined both on Windows and on OS/2:
```suggestion
#if defined(Q_OS_DOSLIKE)
```

dmik commented on src/libs/utils/hostosinfo.h:
```diff
@@ -62,6 +64,7 @@ class QTCREATOR_UTILS_EXPORT HostOsInfo
     static HostArchitecture hostArchitecture();
 
     static constexpr bool isWindowsHost() { return hostOs() == OsTypeWindows; }
+    static constexpr bool isOs2Host() { return hostOs() == OsTypeOs2; }
```
I would also add this:
```suggestion
    static constexpr bool isOs2Host() { return hostOs() == OsTypeOs2; }
    static constexpr bool isDosLikeHost() { return hostOs() == OsTypeWindows || hostOs() == OsTypeOs2; }
```
and use it everywhere instead of `isWindowsHost() || isOs2Host()`.

dmik commented on src/libs/utils/consoleprocess.cpp:
```diff
@@ -364,7 +368,7 @@ void ConsoleProcess::setAbortOnMetaChars(bool abort)
 qint64 ConsoleProcess::applicationMainThreadID() const
 {
-    if (HostOsInfo::isWindowsHost())
+    if (HostOsInfo::isWindowsHost() || HostOsInfo::isOs2Host())
```
```suggestion
    if (HostOsInfo::isDosLikeHost())
```
This requires a suggestion listed below.

dmik commented on src/plugins/coreplugin/dialogs/externaltoolconfig.cpp:
```diff
@@ -619,7 +619,7 @@ void ExternalToolConfig::updateEffectiveArguments()
 void ExternalToolConfig::editEnvironmentChanges()
 {
-    const QString placeholderText = Utils::HostOsInfo::isWindowsHost()
+    const QString placeholderText = Utils::HostOsInfo::isWindowsHost() || Utils::HostOsInfo::isOs2Host()
```
```suggestion
    const QString placeholderText = Utils::HostOsInfo::isDosLikeHost()
```
See previous commit.

dmik commented on src/plugins/projectexplorer/abi.cpp:
```diff
@@ -418,6 +419,23 @@ static Abis abiOf(const QByteArray &data)
         result.append(Abi(Abi::AsmJsArchitecture, Abi::UnknownOS, Abi::UnknownFlavor,
                           Abi::EmscriptenFormat, 32));
     } else if (data.size() >= 64){
+
```
Forgotten empty line?

ydario replied:
I like empty lines around code blocks.

dmik commented on src/plugins/projectexplorer/abi.cpp:
```diff
@@ -433,6 +451,9 @@ static Abis abiOf(const QByteArray &data)
         if (getUint8(data, pePos) == 'P' && getUint8(data, pePos + 1) == 'E'
             && getUint8(data, pePos + 2) == 0 && getUint8(data, pePos + 3) == 0)
             result = parseCoffHeader(data.mid(pePos + 4));
+
```
One more empty line :)

ydario replied:
I like empty lines around code blocks. :-)

dmik commented on src/plugins/projectexplorer/abi.cpp:
```diff
@@ -433,6 +451,9 @@ static Abis abiOf(const QByteArray &data)
             result = parseCoffHeader(data.mid(pePos + 4));
+
+        }
+
```
And one more.

ydario replied:
I like empty lines around code blocks. :-)

dmik commented on src/plugins/projectexplorer/devicesupport/localprocesslist.cpp:
```diff
@@ -86,7 +86,7 @@ QList<DeviceProcessItem> LocalProcessList::getLocalProcesses()
 #endif //Q_OS_WIN
 
-#ifdef Q_OS_UNIX
+#if defined(Q_OS_UNIX) || defined(Q_OS_OS2)
```
There is Q_OS_UNIXLIKE:
```suggestion
#if defined(Q_OS_UNIXLIKE)
```

dmik commented on src/libs/utils/hostosinfo.h:
```diff
@@ -62,6 +64,7 @@ class QTCREATOR_UTILS_EXPORT HostOsInfo
+    static constexpr bool isOs2Host() { return hostOs() == OsTypeOs2; }
     static constexpr bool isLinuxHost() { return hostOs() == OsTypeLinux; }
     static constexpr bool isMacHost() { return hostOs() == OsTypeMac; }
     static constexpr bool isAnyUnixHost()
```
```suggestion
    static constexpr bool isUnixLikeHost() { return isAnyUnixHost() || hostOs() == OsTypeOs2; }
    static constexpr bool isAnyUnixHost()
```

dmik commented on src/shared/proparser/ioutils.cpp:
```diff
@@ -68,7 +68,7 @@ bool IoUtils::isRelativePath(const QString &path)
         return false;
 #endif
-#ifdef Q_OS_WIN
+#if defined(Q_OS_WIN) || defined(Q_OS_OS2)
```
```suggestion
#if defined(Q_OS_DOSLIKE)
```

dmik commented on src/shared/proparser/ioutils.cpp:
```diff
@@ -219,7 +219,7 @@ static QString windowsErrorCode()
 bool IoUtils::touchFile(const QString &targetFileName, const QString &referenceFileName, QString *errorString)
 {
-#  ifdef Q_OS_UNIX
+#  if defined(Q_OS_UNIX) || defined(Q_OS_OS2)
```
```suggestion
#  if defined(Q_OS_UNIXLIKE)
```

dmik commented on src/app/main.cpp:
```diff
@@ -119,15 +119,15 @@ static inline QString toHtml(const QString &t)
 static void displayHelpText(const QString &t)
 {
-    if (Utils::HostOsInfo::isWindowsHost() && qApp)
+    if ((Utils::HostOsInfo::isWindowsHost() || Utils::HostOsInfo::isOs2Host()) && qApp)
```
```suggestion
    if ((Utils::HostOsInfo::isDosLikeHost()) && qApp)
```

dmik commented on src/libs/tracing/timelineitemsrenderpass.cpp:
```diff
@@ -29,6 +29,16 @@
 #include <QtAlgorithms>
 
+#ifdef __OS2__
+// YD hack
```
Please be more specific what this hack is for and why. And no need to put your initials here as they are in the commit anyway :)

ydario replied:
this was not intended to be committed so early...

dmik commented on qtcreator.pri:
```diff
@@ -131,7 +131,10 @@ osx {
 
     IDE_LIBRARY_PATH = $$IDE_OUTPUT_PATH/$$IDE_LIBRARY_BASENAME/qtcreator
-    IDE_PLUGIN_PATH  = $$IDE_LIBRARY_PATH/plugins
+    os2: \
```
Hmm, why putting plugins into /bin/plugins instead of /plugins as on all other systems?

ydario replied:
in my manjaro build the plugins are under bin

dmik commented on src/libs/sqlite/sqlite-lib.pri:
```diff
@@ -8,7 +8,8 @@ INCLUDEPATH += $$PWD
 unix:!bsd: LIBS += -ldl
 
-include(../3rdparty/sqlite/sqlite.pri)
+!os2:include(../3rdparty/sqlite/sqlite.pri)
+os2: LIBS += -lsqlite3
```
Isn't there a Qt Creator configure option for that? Anyway, if there is none, here's the better version:
```suggestion
os2: LIBS += -lsqlite3
else:include(../3rdparty/sqlite/sqlite.pri)
```

ydario replied:
I don't see one.

dmik commented on src/qtcreatorlibrary.pri:
```diff
@@ -7,7 +7,7 @@ include(../qtcreator.pri)
 isEmpty(PRECOMPILED_HEADER):PRECOMPILED_HEADER = $$PWD/shared/qtcreator_pch.h
 
-win32|os2 {
+win32 {
```
Why is OS/2 removed?

ydario replied:
because dlltarget seems not be supported for os2, I see it only in win32 qmake code.

dmik commented on src/qtcreatorlibrary.pri:
```diff
@@ -28,9 +28,12 @@ CONFIG += shared dll
 contains(QT_CONFIG, reduce_exports):CONFIG += hide_symbols
 
-win32|os2 {
+win32 {
```
See above.

ydario replied:
because dlltarget seems not be supported for os2, I see it only in win32 qmake code.

dmik reviewed (changes requested):
I think commits 1c20797 and c59343e can be joined, can't they? Also, c59343e refers to os2utils.cpp and os2utils.h but they are missing from the commits.

Also, if you take Qt Creator code from the old OS/2 repository, please refer the respective commits (with a full GitHub URL) in the commit message — this might greatly help later when doing further patches and/or bug fixing!

Also, commit d91a469 shouldn't actually be there. An IDE of choice is not part of the project, it's part of the developer's environment that shouldn't be committed.

ydario committed:
Update src/shared/qtlockedfile/qtlockedfile.pri
Update src/libs/sqlite/sqlite-lib.pri
Update qtcreator.pro
Update src/plugins/help/qlitehtml/qlitehtml.pri
Update qtcreatordata.pri

Co-Authored-By: Dmitriy Kuminov <coding@dmik.org>
